File: golang/2735_min_cost/min_cost.go
/*
 * @Date: 2023-12-28
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-12-28
 * @FilePath: /algorithm/golang/2735_min_cost/min_cost.go
 */

// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func minCost(nums []int, x int) int64 {
	sum := func(arr []int) int64 {
		ans := int64(0)
		for _, a := range arr {
			ans += int64(a)
		}
		return ans
	}
	n := len(nums)
	f := make([]int, n)
	copy(f, nums)
	ans := sum(f)
	for k := 1; k < n; k++ {
		for i := 0; i < n; i++ {
			f[i] = min(f[i], nums[(i+k)%n])
		}
		ans = min(ans, int64(k)*int64(x)+sum(f))
	}
	return ans
}

func main() {
	tests := []struct {
		nums []int
		x    int
		ans  int64
	}{
		{[]int{20, 1, 15}, 5, 13},
		{[]int{1, 2, 3}, 4, 6},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, minCost(test.nums, test.x), index)
	}
}

File: cpp/2735_min_cost/min_cost.cpp
/*
 * @Date: 2023-12-28
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-12-28
 * @FilePath: /algorithm/cpp/2735_min_cost/min_cost.cpp
 */

#include <cassert>
#include <numeric>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  long long minCost(vector<int>& nums, int x) {
    int n = nums.size();
    vector<int> f(nums);
    long long ans = accumulate(f.begin(), f.end(), 0LL);
    for (int k = 1; k < n; ++k) {
      for (int i = 0; i < n; ++i) {
        f[i] = min(f[i], nums[(i + k) % n]);
      }
      ans = min(ans, static_cast<long long>(k) * x + accumulate(f.begin(), f.end(), 0LL));
    }
    return ans;
  }
};

int main() {
  vector<tuple<vector<int>, int, long long>> tests{
      {{20, 1, 15}, 5, 13},
      {{1, 2, 3}, 4, 6},
  };

  for (auto& [nums, x, ans] : tests) {
    assert(Solution().minCost(nums, x) == ans);
  }
}